### PR TITLE
[WebAssembly] Implement all f16x8 binary instructions.

### PR DESCRIPTION
**Pull request by brendandahl**
This reuses most of the code that was created for f32x4 and f64x2 binary instructions and tries to follow how they were implemented.

add/sub/mul/div - use regular LL instructions
min/max - use the minimum/maximum intrinsic, and also have builtins
pmin/pmax - use the wasm.pmax/pmin intrinsics and also have builtins

Specified at:
https://github.com/WebAssembly/half-precision/blob/29a9b9462c9285d4ccc1a5dc39214ddfd1892658/proposals/half-precision/Overview.md

### CODE REVIEW

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td [hunk 1]
```diff
@@ -1199,6 +1213,7 @@ def : Pat<(v2f64 (froundeven (v2f64 V128:$src))), (NEAREST_F64x2 V128:$src)>;
 multiclass SIMDBinaryFP<SDPatternOperator node, string name, bits<32> baseInst> {
   defm "" : SIMDBinary<F32x4, node, name, baseInst>;
   defm "" : SIMDBinary<F64x2, node, name, !add(baseInst, 12)>;
+  defm "" : SIMDBinary<F16x8, node, name, !add(baseInst, 80), [HasHalfPrecision]>;
```
Comment (aheejin) [inline llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td:1216]: I understand why it's added, and I wish we can multi-inherit from `SIMDBinary` and `HALF_PRECISION_I`, but I'm not sure if we can do it... (Can we?)

I'm not strongly opinionated about it and it's basically just a matter of preference, but how about adding a `multiclass` like `HalfPrecisionBinary` or something that inherits from `SIMDBinary`?

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td [hunk 2]
```diff
@@ -152,6 +153,18 @@ def F64x2 : Vec {
...
+ let lane_vt = f32;
+ let lane_rc = F32;
+ let lane_bits = 16;
+ let lane_idx = LaneIdx8;
+ let lane_load = int_wasm_loadf16_f32;
+ let splat = PatFrag<(ops node:$x), (v8f16 (splat_vector (f16 $x)))>;
+ let prefix = "f16x8";
+}
+
 defvar AllVecs = [I8x16, I16x8, I32x4, I64x2, F32x4, F64x2];
```
Comment (aheejin) [inline llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td:168]: Given that we can't add `F16x8` here, it's not "all vectors" anymore... Should we rename it to something? If so, to what?

**Review by aheejin — COMMENTED** [review]
Now you have commit access!

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td (hunk 2 above)
Comment (brendandahl) [reply llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td:169]: I hope to include F16x8 here when we better support it and the regular patterns work for it. I've added a comment for now, but can change the name if wanted.

File: llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td (hunk 1 above)
Comment (brendandahl) [reply llvm/lib/Target/WebAssembly/WebAssemblyInstrSIMD.td]: I ended up adding `HalfPrecisionBinary`. I was hoping there was some way I could pass a multiclass id as a parameter so i could then pass in `SIMD_I` or `HALF_PRECISION_I` as an argument, but I couldn't figure out a way to make that work.

**Review by aheejin — APPROVED** [review]

**Pull request merged**